### ISSUE TITLE
Not working on WSL (debian) on Windoes 11

### ISSUE THREAD

**Issue by stevef1uk** [exo-explore/exo#536]
Just wondering if this is expect to work on Windows?

I wanted a NVIDIA GPU and it came in a gaming PC (470 Super)

I quickly realised this  project wouldn't work natively on Windows so I tried it on WSL using debian: 

Linux W11 5.15.167.4-microsoft-standard-WSL2 #1 SMP Tue Nov 5 00:21:55 UTC 2024 x86_64 GNU/Linux

After a LOT of time trying to get the NVIDIA cuda packages installed I eventually managed:

```
 python --version
Python 3.12.2

 nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2024 NVIDIA Corporation
Built on Tue_Oct_29_23:50:19_PDT_2024
Cuda compilation tools, release 12.6, V12.6.85
Build cuda_12.6.r12.6/compiler.35059454_0


nvidia-smi
Fri Dec  6 19:43:15 2024
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 560.35.05              Driver Version: 556.12         CUDA Version: 12.5     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 4070 ...    On  |   00000000:01:00.0  On |                  N/A |
| 30%   21C    P8              7W /  220W |     569MiB /  12282MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+

+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A       193      G   /Xwayland                                   N/A      |
+-----------------------------------------------------------------------------------------+


Resulting weights from Softmax:
0.0000000 0.0000000 0.0000000 1.0000000 0.0000000 0.0000714 0.0000000 0.0000000 0.0000000 0.0000000
Loading image data/five_28x28.pgm
Performing forward propagation ...
Resulting weights from Softmax:
0.0000000 0.0000008 0.0000000 0.0000002 0.0000000 1.0000000 0.0000154 0.0000000 0.0000012 0.0000006

Result of classification: 1 3 5

Test passed!
```

I found an issue with the NVIDIA driver as the line in exo/topology/device_capabilities.py:

I changed:

```
 #gpu_raw_name = pynvml.nvmlDeviceGetName(handle).upper()
 gpu_raw_name = "NVIDIA GEFORCE RTX 4070 SUPER"
```

Then exo runs and initially looks good just with this single node. However, with the first small model a simple 'hi' reslted in a generated response of 
```
HowQuestion of you you you you you you you you you you you you you you you you you you you you you you you you you you
```

and apalling performance before exo crashed.

Any ideas

**Comment by wbic16** [comment]
+1 I've also had trouble getting exo to work under WSL2 + Windows 11. I ended up dual-booting Ubuntu 24.04 LTS to move past it for now.

**Comment by stevef1uk** [comment]
> +1 I've also had trouble getting exo to work under WSL2 + Windows 11. I ended up dual-booting Ubuntu 24.04 LTS to move past it for now.

Yes, after I raised my issues I found your issue had been raised. I am a Mac / linux guy so that would be an option, but I was going to install linux I would replace Windows 11. I probably won't do that just to get this project working on my brand new PC ;-)

**Comment by Fahad16301139** [comment]
Please I need a littile bit of help. I am trying to run exo on WSL as well. when i run just the exo command it shows me available commands. but when I try to run the command "exo run llama-3.2-3b" I get "error: unknown command "run" for "exo"

Did you mean this?
runstatus" error. Please help.

**Comment by mandolf0** [comment]
What port should I change this to in order for a Linux machine to discover the WSL exo on a windows machine?
`netsh interface portproxy add v4tov4 listenport=5678 listenaddress=0.0.0.0 connectport=5678 connectaddress=172.19.176.16`
Where 172.19.176.16 comes from `ip addr show eth0` on the WSL terminal.

Coming from this [video](https://youtu.be/yCK3easuYm4?si=ZLKNPfq3N-tgi5ON&t=429): 

TIA

**Comment by diGiCTF** [comment]
I would really like to see exo work with WSL as well. This would be a great workaround for Windows users. I can't get it to work and it's frustrating.

**Comment by murflaw7424** [comment]
I think I figured out how to get this to work. I am on Windows 10 Pro, with WSL Ubuntu, and I was able to get it to work, the following way. Not easy, or quick, but hopefully it works for you and someone else. I am not a professional coder, and more of a tinkerer/hobbyist, so my apologies if my code steps do not make sense, or are not formatted correctly.

### 1. Identify your WSL2 IP Address (open another terminal instance of wsl, or close exo):

In your WSL Ubuntu Terminal, run:

`hostname -I`

It might output something like:

`172.24.255.21 172.17.0.1 172.18.0.1`

Use only the first IP (e.g., 172.24.255.21), since the others are likely docker containers or another service you have (I had 3).

### 2. Configure Windows Port Forwarding:

On Windows, **_open PowerShell as Administrator_** and run:

Replace <WSL-IP> with the IP address you identified above:

```
netsh interface portproxy delete v4tov4 listenport=52415 listenaddress=0.0.0.0
netsh interface portproxy add v4tov4 listenport=52415 listenaddress=0.0.0.0 connectport=52415 connectaddress=<WSL-IP>
```

For example:

`netsh interface portproxy add v4tov4 listenport=52415 listenaddress=0.0.0.0 connectport=52415 connectaddress=172.24.255.21`

### 3. Add Firewall Exception (required once):

Still in **_Admin PowerShell_**, run:

`New-NetFirewallRule -DisplayName 'WSL2 Exo Port' -Direction Inbound -LocalPort 52415 -Action Allow -Protocol TCP`

Then Run:

`ssh -N -L 52415:localhost:52415 $(wsl hostname -I).Split()[0]`

**_It may prompt you for a password, which should be your sudo password in WSL. However, I could not get the password to work, but my fix would not work without running this command._**

### 4. Access Exo from Windows:

Open your Windows browser and visit:

`http://localhost:52415`

✅ You're done! Your Exo interface running inside WSL2 is now accessible from Windows.

⚠️ If your WSL IP changes after rebooting WSL:

Repeat step 1 to identify the new IP and re-run step 2 to update the forwarding rule:

```
netsh interface portproxy delete v4tov4 listenport=52415 listenaddress=0.0.0.0
netsh interface portproxy add v4tov4 listenport=52415 listenaddress=0.0.0.0 connectport=52415 connectaddress=<NEW-WSL-IP>
```

After 3 reboots, my WSL IP has not changed. Now you'll have easy access!

**Comment by LGDHuaOPER** [comment]
I have encountered issues between my Macbook and Windows 10 WSL-Ubuntu 22, these two machines are on the same local network. And I forwarded ports 5678 and 52415 on Windows10 using `netsh interface portproxy add v4tov4` and also opened firewall inbound rules for 5678 (UDP) and 52415 (TCP). The result is still that these two machines cannot discover each other. What is the reason? Thanks!

**Comment by ro-mak** [comment]
> I have encountered issues between my Macbook and Windows 10 WSL-Ubuntu 22, these two machines are on the same local network. And I forwarded ports 5678 and 52415 on Windows10 using `netsh interface portproxy add v4tov4` and also opened firewall inbound rules for 5678 (UDP) and 52415 (TCP). The result is still that these two machines cannot discover each other. What is the reason? Thanks!

I managed to connect them via manual topology
and you should also set --node-port and open it that is required for grpc server
